### PR TITLE
Tf/contrastive prediction

### PR DESCRIPTION
**Pull request by marcromeyn**
**Blocked by:**
- #589

### Goals :soccer:
- New implementation of the contrastive task used to train retrieval models
- Fix saving/loading two-tower model: #498

### Implementation Details :construction:

- Register metrics and makes sure TopKMetricsAggregator is serializable  
- Wrap the top-k metrics provided to the model in a  `TopKMetricsAggregator` instance to make sure top-k extraction and label relevant counts are computed once for all metrics. 
- Add a base-class `ContrastivePredictionBlock` . 
- Add `ContrastiveDotProduct` that replaces the ItemRetrievalScorer. This new implementation does not rely on the context layer and has direct access to the query embeddings. 
-  Add a new`DotProductCategoricalPrediction` for prediction tasks including negative sampling (will replace the `ItemRetrievalTask`)
---> (**Note**: This new implementation does not support the TopkIndex evaluation).
---> The motivation is to decouple the contrastive learning task of the RetrievalModel from the evaluation of the top-k recommender. 
---> So the supported evaluation strategy is to use negative sampling, similar to training. 

### Testing Details :mag:

- Add unit test of saving/loading RetrievalModel with the new `DotProductCategoricalPrediction` block. 
- Add unit tests for `ContrastivePredictionBlock`.
- Add a unit test for specifying the negative sampling strategy in the `compile` method.

### CODE REVIEW

**Comment by viswa-nvidia** [comment]
Moved to arbitration. Not clear how this maps to the r[oadmap](https://github.com/orgs/NVIDIA-Merlin/projects/14/views/4) . @marcromeyn , please clarify.

File: merlin/models/tf/predictions/sampling/in_batch.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from typing import Optional
+
+import tensorflow as tf
+
+from merlin.models.tf.predictions.sampling.base import Items, ItemSampler
+
+
+@ItemSampler.registry.register("in-batch")
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class InBatchSampler(ItemSampler):
```
Comment (oliverholworthy) [inline merlin/models/tf/predictions/sampling/in_batch.py:25]: What is different about this compared to the `InBatchSampler` that exists in `tf.blocks.sampling.in_batch` ?

Comment (sararb) [reply merlin/models/tf/predictions/sampling/in_batch.py:25]: The objective is to move the sampling under the predictions folder as this is used by the new refactored prediction blocks. There are also some enhancements of the sampling blocks (base and InBatchSampler) to improve the API, which are: 

-  Add a registry for sampling strategies 
-  The class arguments are _last_batch_items_embeddings and _last_batch_items_metadata are wrapped in  a new NamedTuple class `Items`.  This new class will replace replace `EmbeddingWithMetadata` class. 
- Add `get_config` method to `InBatchSampler` for serialization. 

I think we could keep the old sampling strategies until we update the High-level API with the new prediction blocks and the old ItemRetrievalTask / ItemRetrievalScorer gets dropped.

File: merlin/models/tf/predictions/dot_product.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+            post=post,
+            logits_temperature=logits_temperature,
+            **kwargs,
+        )
+
+    def compile(self, negative_sampling=None, downscore_false_negatives=False):
+        self.prediction_with_negatives.negative_sampling = negative_sampling
+        self.prediction_with_negatives.downscore_false_negatives = downscore_false_negatives
+
+    # TODO
```
Comment (oliverholworthy) [inline merlin/models/tf/predictions/dot_product.py:118]: If this TODO is still relevant, can you add more context of what it refers to (given that the method seems to have an implementation

File: merlin/models/tf/predictions/sampling/in_batch.py (hunk 1 above)
Comment (oliverholworthy) [reply merlin/models/tf/predictions/sampling/in_batch.py:25]: We can follow-up on this in another PR. I'm wondering if the behavior of this class is the same from an interface/api perspective - will we be able to replace uses of the old one without any changes to the surrounding code.

File: merlin/models/tf/predictions/sampling/in_batch.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        return items
+
+    def sample(self) -> Items:
+        return self._last_batch
+
+    def get_config(self):
+        config = super().get_config()
+        config["batch_size"] = self._batch_size
+
+        # TODO: This is a side-effect, could this lead to problems?
```
Comment (oliverholworthy) [inline merlin/models/tf/predictions/sampling/in_batch.py:86]: can you explain why is this line necessary? something to do with resetting the state?

File: merlin/models/tf/predictions/dot_product.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+from merlin.models.tf.predictions.base import ContrastivePredictionBlock
+from merlin.models.tf.predictions.sampling.base import Items, ItemSampler, ItemSamplersType
+from merlin.models.tf.typing import TabularData
+from merlin.models.tf.utils import tf_utils
+from merlin.models.tf.utils.tf_utils import call_layer, rescore_false_negatives
+from merlin.models.utils import schema_utils
+from merlin.models.utils.constants import MIN_FLOAT
+from merlin.schema import Schema, Tags
+
+LOG = logging.getLogger("merlin_models")
```
Comment (oliverholworthy) [inline merlin/models/tf/predictions/dot_product.py:18]: 🗒️  this doesn't need to be addressed in this PR. This is an observation that we should address the inconsistency we have in the name we used inside `getLogger` across the repo:

https://github.com/NVIDIA-Merlin/models/search?p=1&q=getLogger

File: merlin/models/tf/predictions/sampling/base.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+import abc
```
Comment (oliverholworthy) [inline merlin/models/tf/predictions/sampling/base.py:1]: missing license from the top of this file, + some of the other new files added

File: merlin/models/tf/predictions/dot_product.py (hunk 4 above)
Comment (sararb) [reply merlin/models/tf/predictions/dot_product.py:18]: Thanks for the catch! I just opened a new issue for tracking

File: merlin/models/tf/predictions/sampling/base.py (hunk 5 above)
Comment (sararb) [reply merlin/models/tf/predictions/sampling/base.py:16]: I have just updated the files with the missing license

File: merlin/models/tf/predictions/sampling/in_batch.py (hunk 3 above)
Comment (sararb) [reply merlin/models/tf/predictions/sampling/in_batch.py:107]: @marcromeyn added this code line, so I kept it until he returned to find out why it is needed.

File: merlin/models/tf/predictions/dot_product.py (hunk 2 above)
Comment (sararb) [reply merlin/models/tf/predictions/dot_product.py]: Thanks for the catch!  I just removed the # TODO as the `add_sampler` is now implemented and tested.

**Review by oliverholworthy — APPROVED** [review]

File: merlin/models/tf/predictions/sampling/in_batch.py (hunk 1 above)
Comment (sararb) [reply merlin/models/tf/predictions/sampling/in_batch.py]: Thanks, that's a good point! Once all tasks defined in this [ticket](https://github.com/NVIDIA-Merlin/Merlin/issues/464) are done, the next step is to do the benchmarking of the retrieval/ranking models and then we change the high-level API. The main constraint when implementing these tasks is to keep the user-facing API the same. 

But in order to simplify things, the item retrieval model will be converted into a Top-k recommender and Top-k evaluation will be implemented differently. The overall objective is to build a `ItemRecommenderModel` class that supports the base methods `predict()` for evaluation and `.save()`  for serialization. [Here](https://github.com/NVIDIA-Merlin/models/compare/main...tf/retrieval-models) is a draft code that summarizes what Marc and I discussed.
